src/main/java/smp/RankingMatrix.java

```diff
@@ -0,0 +1,466 @@
+            increasingArray[increasingIndex] = increasingIndex;
+        }
+
+        int[] permutatedArray;
+        int[] matchesByColumn;
+
+        do {
+//      Compute M(R_i)
+            permutatedArray = randomPermutation(increasingArray, m_numberOfColumns);
+            matchesByColumn = getMatchesbyColumn(permutatedArray);
+
+            updateCurrentMatching(permutatedArray);
+
+
+            int iterationCounter = 0;
+
+            do {
+
+                HashMap<String, LinkedList<MatrixCoordinate>> unstablePairsPerRow = GetUnstablePairsPerRow(permutatedArray, matchesByColumn);
+
+                //TODO: Remove
+                System.out.println("Found " + "unstable pairs");
+                for (String row : unstablePairsPerRow.keySet()) {
+                    for (MatrixCoordinate pair : unstablePairsPerRow.get(row)) {
+                        System.out.println(pair.toString());
+                    }
+                }
+
+                LinkedList<MatrixCoordinate> nm1_pairs = getNm1_Pairs(unstablePairsPerRow);
+
+                LinkedList<MatrixCoordinate> nm2_pairs = getNm2_Pairs(nm1_pairs, matchesByColumn, permutatedArray);
+
+                LinkedList<MatrixCoordinate> nm_pairs = new LinkedList<MatrixCoordinate>();
+                nm_pairs.addAll(nm1_pairs);
+                nm_pairs.addAll(nm2_pairs);
+
+                //TODO:remove print
+                System.out.println("nm_pairs");
+                for (MatrixCoordinate nm_pair : nm_pairs) {
+                    //TODO: Remove print
+                    System.out.println(nm_pair.toString());
+                }
+
+                permutatedArray = iterateMatching(permutatedArray, nm1_pairs);
+                matchesByColumn = getMatchesbyColumn(permutatedArray);
+                updateCurrentMatching(permutatedArray);
+            }while(!GetUnstablePairsPerRow(permutatedArray, matchesByColumn).isEmpty() && (++iterationCounter<m_numberOfColumns*3));
+        }while(!GetUnstablePairsPerRow(permutatedArray, matchesByColumn).isEmpty());
+    }
+
+    private int[] iterateMatching(int[] previousMatching, LinkedList<MatrixCoordinate> nm_pairs)
+    {
+        for (MatrixCoordinate new_pair: nm_pairs)
+        {
+            previousMatching[new_pair.getI()]= new_pair.getJ();
+        }
+
+        return previousMatching;
+    }
+
+    private HashMap<String, LinkedList<MatrixCoordinate>> GetUnstablePairsPerRow(int[] permutatedArray, int[] tranposedIndices)
+    {
+        HashMap<String, LinkedList<MatrixCoordinate>> unstablePairsPerRow = new HashMap<String, LinkedList<MatrixCoordinate>>();
+
+        // Unstable pair detection.
+        // TODO: parallelize
+        for (int i = 0; i < m_numberOfRows; ++i) {
+            for (int j = 0; j < m_numberOfColumns; ++j) {
+                //a_i,j is unstable if a_i,j^L < M(R_i)^L and a_i,j^R < M(C_j)^R
+                if ((m_rankingMatrix[i][j].getLeftValue() < m_rankingMatrix[i][permutatedArray[i]].getLeftValue()) &&
```

Comment:
The code for this nested fors can be parallelized, the order of the output LinkedList does not matter

Reply:
Ok, I will start working on it now.